internal/tofu/context_plugins.go

```diff
@@ -70,6 +70,11 @@ func (cp *contextPlugins) ProviderSchema(addr addrs.Provider) (providers.Provide
 	// This cache is only written by the provider client, and transparently
 	// used by GetProviderSchema, but we check it here because at this point we
 	// may be able to avoid spinning up the provider instance at all.
+	//
+	// It's worth noting that ServerCapabilities.GetProviderSchemaOptional is ignored here.
+	// That is because we're checking *prior* to the provider's instantiation.
+	// GetProviderSchemaOptional only says that *if we instantiate a provider*,
+	// then we need to run the get schema call at least once.
```

Comment:
Sorry, but I don't get the point of this comment. AFAIU `contextPlugins.ProviderSchema` does not keep track of *prior* initialisation(s) for decision on whether `GetProviderSchemaOptional` value is relevant for cache. On that basis, I find this comment confusing here.

Reply:
@kislerdm The provider is getting instantiated a few lines later, and this is an early return.

So this is clarifying that it's ok to read this even if GetProviderSchemaOptional is false, because the provider hasn't yet been instantiated (so doesn't require setup) and we have the schema leftover from a previous provider instantiation (one that is possibly now no more running).